---
# Pandas Data Analysis

https://pandas.pydata.org/docs/getting_started/intro_tutorials/index.html

https://pandas.pydata.org/docs/user_guide/index.html

In [ ]:
import numpy as np
import pandas as pd

---
## The `nycflights13::flights` data

The [Python nycflights13](https://pypi.org/project/nycflights13/) data package provides the same data as the [R nycflights13](https://cran.r-project.org/web/packages/nycflights13/index.html) package.

![](http://r4ds.had.co.nz/diagrams/relational-nycflights.png)

In [ ]:
# install the package
!pip install nycflights13

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.7/8.7 MB 42.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for nycflights13: filename=nycflights13-0.0.3-py3-none-any.whl size=8732741 sha256=de259df7a7827c7a62219164045ee489c1e91d13627951ca42a1c71175ed1824
  Stored in directory: /root/.cache/pip/wheels/b6/27/3d/46507f17840b411b66f10620728643140d0e6fa037df0fc9d8
Successfully built nycflights13


In [ ]:
# load the `flights` table
from nycflights13 import flights
print(type(flights))
flights.head()

<class 'pandas.core.frame.DataFrame'>


year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute             time_hour  
0      1400     5      15  2013-01-01T10:00:00Z  
1      1416     5      29  2013-01-01T10:00:00Z  
2      1089     5      40  2013-01-01T10:00:00Z  
3      1576     5      45  2013-01-01T10:00:00Z  
4       762     6       0  2013-01-01T11:00:00Z

In [ ]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 336776 entries, 0 to 336775
Data columns (total 19 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   year            336776 non-null  int64  
 1   month           336776 non-null  int64  
 2   day             336776 non-null  int64  
 3   dep_time        328521 non-null  float64
 4   sched_dep_time  336776 non-null  int64  
 5   dep_delay       328521 non-null  float64
 6   arr_time        328063 non-null  float64
 7   sched_arr_time  336776 non-null  int64  
 8   arr_delay       327346 non-null  float64
 9   carrier         336776 non-null  object 
 10  flight          336776 non-null  int64  
 11  tailnum         334264 non-null  object 
 12  origin          336776 non-null  object 
 13  dest            336776 non-null  object 
 14  air_time        327346 non-null  float64
 15  distance        336776 non-null  int64  
 16  hour            336776 non-null  int64  
 17  minute    

This `flights` dataframe contains all 336,776 flights that departed from New York City in 2013.

---
## Missing value `NaN`

There are some missing values in the flights data set which were caused by canceled flights. 
We can clean the flights data by removing flight record that has missing values in `dep_delay` or `arr_delay`.

In [ ]:
# use notnull() or notna()
flights[flights['dep_delay'].notna() & flights['arr_delay'].notna()]

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336765  2013      9   30    2240.0            2245       -5.0    2334.0   
336766  2013      9   30    2240.0            2250      -10.0    2347.0   
336767  2013      9   30    2241.0            2246       -5.0    2345.0   
336768  2013      9   30    2307.0            2255       12.0    2359.0   
336769  2013      9   30    2349.0            2359      -10.0     325.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
3                 1022      -18.0      B6     725  N804JB    JFK  BQN   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
...                ...        ...     ...     ...     ...    ...  ...   
336765            2351      -17.0      B6    1816  N354JB    JFK  SYR   
336766               7      -20.0      B6    2002  N281JB    JFK  BUF   
336767               1      -16.0      B6     486  N346JB    JFK  ROC   
336768            2358        1.0      B6     718  N565JB    JFK  BOS   
336769             350      -25.0      B6     745  N516JB    JFK  PSE   

        air_time  distance  hour  minute             time_hour  
0          227.0      1400     5      15  2013-01-01T10:00:00Z  
1          227.0      1416     5      29  2013-01-01T10:00:00Z  
2          160.0      1089     5      40  2013-01-01T10:00:00Z  
3          183.0      1576     5      45  2013-01-01T10:00:00Z  
4          116.0       762     6       0  2013-01-01T11:00:00Z  
...          ...       ...   ...     ...                   ...  
336765      41.0       209    22      45  2013-10-01T02:00:00Z  
336766      52.0       301    22      50  2013-10-01T02:00:00Z  
336767      47.0       264    22      46  2013-10-01T02:00:00Z  
336768      33.0       187    22      55  2013-10-01T02:00:00Z  
336769     196.0      1617    23      59  2013-10-01T03:00:00Z  

[327346 rows x 19 columns]

In [ ]:
# use dropna()
flights.dropna(subset = ['dep_delay', 'arr_delay'])

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336765  2013      9   30    2240.0            2245       -5.0    2334.0   
336766  2013      9   30    2240.0            2250      -10.0    2347.0   
336767  2013      9   30    2241.0            2246       -5.0    2345.0   
336768  2013      9   30    2307.0            2255       12.0    2359.0   
336769  2013      9   30    2349.0            2359      -10.0     325.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
3                 1022      -18.0      B6     725  N804JB    JFK  BQN   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
...                ...        ...     ...     ...     ...    ...  ...   
336765            2351      -17.0      B6    1816  N354JB    JFK  SYR   
336766               7      -20.0      B6    2002  N281JB    JFK  BUF   
336767               1      -16.0      B6     486  N346JB    JFK  ROC   
336768            2358        1.0      B6     718  N565JB    JFK  BOS   
336769             350      -25.0      B6     745  N516JB    JFK  PSE   

        air_time  distance  hour  minute             time_hour  
0          227.0      1400     5      15  2013-01-01T10:00:00Z  
1          227.0      1416     5      29  2013-01-01T10:00:00Z  
2          160.0      1089     5      40  2013-01-01T10:00:00Z  
3          183.0      1576     5      45  2013-01-01T10:00:00Z  
4          116.0       762     6       0  2013-01-01T11:00:00Z  
...          ...       ...   ...     ...                   ...  
336765      41.0       209    22      45  2013-10-01T02:00:00Z  
336766      52.0       301    22      50  2013-10-01T02:00:00Z  
336767      47.0       264    22      46  2013-10-01T02:00:00Z  
336768      33.0       187    22      55  2013-10-01T02:00:00Z  
336769     196.0      1617    23      59  2013-10-01T03:00:00Z  

[327346 rows x 19 columns]

---
## Filter rows based on conditions

- Pandas: boolean index or `query()` https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html

In [ ]:
# indexing
flights[flights['month'] == 1]
#flights[flights.month == 1]

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0      2013      1    1     517.0             515        2.0     830.0   
1      2013      1    1     533.0             529        4.0     850.0   
2      2013      1    1     542.0             540        2.0     923.0   
3      2013      1    1     544.0             545       -1.0    1004.0   
4      2013      1    1     554.0             600       -6.0     812.0   
...     ...    ...  ...       ...             ...        ...       ...   
26999  2013      1   31       NaN            1325        NaN       NaN   
27000  2013      1   31       NaN            1200        NaN       NaN   
27001  2013      1   31       NaN            1410        NaN       NaN   
27002  2013      1   31       NaN            1446        NaN       NaN   
27003  2013      1   31       NaN             625        NaN       NaN   

       sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                 819       11.0      UA    1545  N14228    EWR  IAH   
1                 830       20.0      UA    1714  N24211    LGA  IAH   
2                 850       33.0      AA    1141  N619AA    JFK  MIA   
3                1022      -18.0      B6     725  N804JB    JFK  BQN   
4                 837      -25.0      DL     461  N668DN    LGA  ATL   
...               ...        ...     ...     ...     ...    ...  ...   
26999            1505        NaN      MQ    4475  N730MQ    LGA  RDU   
27000            1430        NaN      MQ    4658  N505MQ    LGA  ATL   
27001            1555        NaN      MQ    4491  N734MQ    LGA  CLE   
27002            1757        NaN      UA     337     NaN    LGA  IAH   
27003             934        NaN      UA    1497     NaN    LGA  IAH   

       air_time  distance  hour  minute             time_hour  
0         227.0      1400     5      15  2013-01-01T10:00:00Z  
1         227.0      1416     5      29  2013-01-01T10:00:00Z  
2         160.0      1089     5      40  2013-01-01T10:00:00Z  
3         183.0      1576     5      45  2013-01-01T10:00:00Z  
4         116.0       762     6       0  2013-01-01T11:00:00Z  
...         ...       ...   ...     ...                   ...  
26999       NaN       431    13      25  2013-01-31T18:00:00Z  
27000       NaN       762    12       0  2013-01-31T17:00:00Z  
27001       NaN       419    14      10  2013-01-31T19:00:00Z  
27002       NaN      1416    14      46  2013-01-31T19:00:00Z  
27003       NaN      1416     6      25  2013-01-31T11:00:00Z  

[27004 rows x 19 columns]

In [ ]:
# indexing by multiple columns
flights[(flights['month'] == 1) & (flights['day'] == 1)]
#flights[(flights.month == 1) & (flights.day == 1)]

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0    2013      1    1     517.0             515        2.0     830.0   
1    2013      1    1     533.0             529        4.0     850.0   
2    2013      1    1     542.0             540        2.0     923.0   
3    2013      1    1     544.0             545       -1.0    1004.0   
4    2013      1    1     554.0             600       -6.0     812.0   
..    ...    ...  ...       ...             ...        ...       ...   
837  2013      1    1    2356.0            2359       -3.0     425.0   
838  2013      1    1       NaN            1630        NaN       NaN   
839  2013      1    1       NaN            1935        NaN       NaN   
840  2013      1    1       NaN            1500        NaN       NaN   
841  2013      1    1       NaN             600        NaN       NaN   

     sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0               819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1               830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2               850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3              1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4               837      -25.0      DL     461  N668DN    LGA  ATL     116.0   
..              ...        ...     ...     ...     ...    ...  ...       ...   
837             437      -12.0      B6     727  N588JB    JFK  BQN     186.0   
838            1815        NaN      EV    4308  N18120    EWR  RDU       NaN   
839            2240        NaN      AA     791  N3EHAA    LGA  DFW       NaN   
840            1825        NaN      AA    1925  N3EVAA    LGA  MIA       NaN   
841             901        NaN      B6     125  N618JB    JFK  FLL       NaN   

     distance  hour  minute             time_hour  
0        1400     5      15  2013-01-01T10:00:00Z  
1        1416     5      29  2013-01-01T10:00:00Z  
2        1089     5      40  2013-01-01T10:00:00Z  
3        1576     5      45  2013-01-01T10:00:00Z  
4         762     6       0  2013-01-01T11:00:00Z  
..        ...   ...     ...                   ...  
837      1576    23      59  2013-01-02T04:00:00Z  
838       416    16      30  2013-01-01T21:00:00Z  
839      1389    19      35  2013-01-02T00:00:00Z  
840      1096    15       0  2013-01-01T20:00:00Z  
841      1069     6       0  2013-01-01T11:00:00Z  

[842 rows x 19 columns]

In [ ]:
# query()
flights.query('month == 1 & day == 1')

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0    2013      1    1     517.0             515        2.0     830.0   
1    2013      1    1     533.0             529        4.0     850.0   
2    2013      1    1     542.0             540        2.0     923.0   
3    2013      1    1     544.0             545       -1.0    1004.0   
4    2013      1    1     554.0             600       -6.0     812.0   
..    ...    ...  ...       ...             ...        ...       ...   
837  2013      1    1    2356.0            2359       -3.0     425.0   
838  2013      1    1       NaN            1630        NaN       NaN   
839  2013      1    1       NaN            1935        NaN       NaN   
840  2013      1    1       NaN            1500        NaN       NaN   
841  2013      1    1       NaN             600        NaN       NaN   

     sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0               819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1               830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2               850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3              1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4               837      -25.0      DL     461  N668DN    LGA  ATL     116.0   
..              ...        ...     ...     ...     ...    ...  ...       ...   
837             437      -12.0      B6     727  N588JB    JFK  BQN     186.0   
838            1815        NaN      EV    4308  N18120    EWR  RDU       NaN   
839            2240        NaN      AA     791  N3EHAA    LGA  DFW       NaN   
840            1825        NaN      AA    1925  N3EVAA    LGA  MIA       NaN   
841             901        NaN      B6     125  N618JB    JFK  FLL       NaN   

     distance  hour  minute             time_hour  
0        1400     5      15  2013-01-01T10:00:00Z  
1        1416     5      29  2013-01-01T10:00:00Z  
2        1089     5      40  2013-01-01T10:00:00Z  
3        1576     5      45  2013-01-01T10:00:00Z  
4         762     6       0  2013-01-01T11:00:00Z  
..        ...   ...     ...                   ...  
837      1576    23      59  2013-01-02T04:00:00Z  
838       416    16      30  2013-01-01T21:00:00Z  
839      1389    19      35  2013-01-02T00:00:00Z  
840      1096    15       0  2013-01-01T20:00:00Z  
841      1069     6       0  2013-01-01T11:00:00Z  

[842 rows x 19 columns]

---
## &diams; Exercise

1. Find the flights that departed in the summer (from July to September)

In [ ]:
# ANSWER
#'month' column is between 7 & 9 (inclusive)
flights[(flights['month'] >= 7) & (flights['month'] <= 9)]

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
250450  2013      7    1       1.0            2029      212.0     236.0   
250451  2013      7    1       2.0            2359        3.0     344.0   
250452  2013      7    1      29.0            2245      104.0     151.0   
250453  2013      7    1      43.0            2130      193.0     322.0   
250454  2013      7    1      44.0            2150      174.0     300.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
250450            2359      157.0      B6     915  N653JB    JFK  SFO   
250451             344        0.0      B6    1503  N805JB    JFK  SJU   
250452               1      110.0      B6     234  N348JB    JFK  BTV   
250453              14      188.0      B6    1371  N794JB    LGA  FLL   
250454             100      120.0      AA     185  N324AA    JFK  LAX   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1634        NaN      9E    3393     NaN    JFK  DCA   
336772            2312        NaN      9E    3525     NaN    LGA  SYR   
336773            1330        NaN      MQ    3461  N535MQ    LGA  BNA   
336774            1344        NaN      MQ    3572  N511MQ    LGA  CLE   
336775            1020        NaN      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute             time_hour  
250450     315.0      2586    20      29  2013-07-02T00:00:00Z  
250451     200.0      1598    23      59  2013-07-02T03:00:00Z  
250452      66.0       266    22      45  2013-07-02T02:00:00Z  
250453     143.0      1076    21      30  2013-07-02T01:00:00Z  
250454     297.0      2475    21      50  2013-07-02T01:00:00Z  
...          ...       ...   ...     ...                   ...  
336771       NaN       213    14      55  2013-09-30T18:00:00Z  
336772       NaN       198    22       0  2013-10-01T02:00:00Z  
336773       NaN       764    12      10  2013-09-30T16:00:00Z  
336774       NaN       419    11      59  2013-09-30T15:00:00Z  
336775       NaN       431     8      40  2013-09-30T12:00:00Z  

[86326 rows x 19 columns]

2. Find the flights that flew to Houston (IAH or HOU)

In [ ]:
# ANSWER
#Here,'|' character is the logical OR operator
flights[flights['dest'].str.contains('IAH|HOU')]

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
32      2013      1    1     623.0             627       -4.0     933.0   
81      2013      1    1     728.0             732       -4.0    1041.0   
89      2013      1    1     739.0             739        0.0    1104.0   
...      ...    ...  ...       ...             ...        ...       ...   
336524  2013      9   30    1729.0            1720        9.0    2001.0   
336527  2013      9   30    1735.0            1715       20.0    2010.0   
336618  2013      9   30    1859.0            1859        0.0    2134.0   
336694  2013      9   30    2015.0            2015        0.0    2244.0   
336737  2013      9   30    2105.0            2106       -1.0    2329.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
32                 932        1.0      UA     496  N459UA    LGA  IAH   
81                1038        3.0      UA     473  N488UA    LGA  IAH   
89                1038       26.0      UA    1479  N37408    EWR  IAH   
...                ...        ...     ...     ...     ...    ...  ...   
336524            2010       -9.0      UA     652  N455UA    EWR  IAH   
336527            2005        5.0      WN    2067  N296WN    EWR  HOU   
336618            2159      -25.0      UA    1128  N14731    LGA  IAH   
336694            2307      -23.0      UA    1545  N17730    EWR  IAH   
336737            2354      -25.0      UA     475  N477UA    EWR  IAH   

        air_time  distance  hour  minute             time_hour  
0          227.0      1400     5      15  2013-01-01T10:00:00Z  
1          227.0      1416     5      29  2013-01-01T10:00:00Z  
32         229.0      1416     6      27  2013-01-01T11:00:00Z  
81         238.0      1416     7      32  2013-01-01T12:00:00Z  
89         249.0      1400     7      39  2013-01-01T12:00:00Z  
...          ...       ...   ...     ...                   ...  
336524     173.0      1400    17      20  2013-09-30T21:00:00Z  
336527     188.0      1411    17      15  2013-09-30T21:00:00Z  
336618     180.0      1416    18      59  2013-09-30T22:00:00Z  
336694     174.0      1400    20      15  2013-10-01T00:00:00Z  
336737     175.0      1400    21       6  2013-10-01T01:00:00Z  

[9313 rows x 19 columns]

---
## Sort records by values

- Pandas: `sort_values()` https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html

In [ ]:
#flights.sort_values('arr_delay')
flights.sort_values('arr_delay', ascending = 0)

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
7072    2013      1    9     641.0             900     1301.0    1242.0   
235778  2013      6   15    1432.0            1935     1137.0    1607.0   
8239    2013      1   10    1121.0            1635     1126.0    1239.0   
327043  2013      9   20    1139.0            1845     1014.0    1457.0   
270376  2013      7   22     845.0            1600     1005.0    1044.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
7072              1530     1272.0      HA      51  N384HA    JFK  HNL   
235778            2120     1127.0      MQ    3535  N504MQ    JFK  CMH   
8239              1810     1109.0      MQ    3695  N517MQ    EWR  ORD   
327043            2210     1007.0      AA     177  N338AA    JFK  SFO   
270376            1815      989.0      MQ    3075  N665MQ    JFK  CVG   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1634        NaN      9E    3393     NaN    JFK  DCA   
336772            2312        NaN      9E    3525     NaN    LGA  SYR   
336773            1330        NaN      MQ    3461  N535MQ    LGA  BNA   
336774            1344        NaN      MQ    3572  N511MQ    LGA  CLE   
336775            1020        NaN      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute             time_hour  
7072       640.0      4983     9       0  2013-01-09T14:00:00Z  
235778      74.0       483    19      35  2013-06-15T23:00:00Z  
8239       111.0       719    16      35  2013-01-10T21:00:00Z  
327043     354.0      2586    18      45  2013-09-20T22:00:00Z  
270376      96.0       589    16       0  2013-07-22T20:00:00Z  
...          ...       ...   ...     ...                   ...  
336771       NaN       213    14      55  2013-09-30T18:00:00Z  
336772       NaN       198    22       0  2013-10-01T02:00:00Z  
336773       NaN       764    12      10  2013-09-30T16:00:00Z  
336774       NaN       419    11      59  2013-09-30T15:00:00Z  
336775       NaN       431     8      40  2013-09-30T12:00:00Z  

[336776 rows x 19 columns]

In [ ]:
# the flight with the longest arrival delay
flights.sort_values('arr_delay', ascending = 0).head(1)

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
7072  2013      1    9     641.0             900     1301.0    1242.0   

      sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
7072            1530     1272.0      HA      51  N384HA    JFK  HNL     640.0   

      distance  hour  minute             time_hour  
7072      4983     9       0  2013-01-09T14:00:00Z

- `n`-largest: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.nlargest.html
- `n`-smallest: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.nsmallest.html

In [ ]:
flights.nlargest(1, 'arr_delay', keep = 'all')

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
7072  2013      1    9     641.0             900     1301.0    1242.0   

      sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
7072            1530     1272.0      HA      51  N384HA    JFK  HNL     640.0   

      distance  hour  minute             time_hour  
7072      4983     9       0  2013-01-09T14:00:00Z

---
## Select columns

- Pandas: indexing by column names or `filter()` https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.filter.html

In [ ]:
# indexing by a list/range of columns
flights[['year', 'month', 'day']]
#flights.loc[:, 'year':'day']

year  month  day
0       2013      1    1
1       2013      1    1
2       2013      1    1
3       2013      1    1
4       2013      1    1
...      ...    ...  ...
336771  2013      9   30
336772  2013      9   30
336773  2013      9   30
336774  2013      9   30
336775  2013      9   30

[336776 rows x 3 columns]

---
## Add a new column
- Pandas: assignment of dataframe['new_col']

In [ ]:
flights['speed'] = flights['distance'] / flights['air_time']
flights

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
3                 1022      -18.0      B6     725  N804JB    JFK  BQN   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1634        NaN      9E    3393     NaN    JFK  DCA   
336772            2312        NaN      9E    3525     NaN    LGA  SYR   
336773            1330        NaN      MQ    3461  N535MQ    LGA  BNA   
336774            1344        NaN      MQ    3572  N511MQ    LGA  CLE   
336775            1020        NaN      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute             time_hour     speed  
0          227.0      1400     5      15  2013-01-01T10:00:00Z  6.167401  
1          227.0      1416     5      29  2013-01-01T10:00:00Z  6.237885  
2          160.0      1089     5      40  2013-01-01T10:00:00Z  6.806250  
3          183.0      1576     5      45  2013-01-01T10:00:00Z  8.612022  
4          116.0       762     6       0  2013-01-01T11:00:00Z  6.568966  
...          ...       ...   ...     ...                   ...       ...  
336771       NaN       213    14      55  2013-09-30T18:00:00Z       NaN  
336772       NaN       198    22       0  2013-10-01T02:00:00Z       NaN  
336773       NaN       764    12      10  2013-09-30T16:00:00Z       NaN  
336774       NaN       419    11      59  2013-09-30T15:00:00Z       NaN  
336775       NaN       431     8      40  2013-09-30T12:00:00Z       NaN  

[336776 rows x 20 columns]

---
## &diams; Exercise
Find the fastest flight (i.e., the flight has the highest speed). Report the flight date, actual departure and arrival time, origin, destination, carrier, and flight number.

The expectED result is DL1499 which departed at 17:09pm from LGA and arrived at 19:23pm to ATL on May 25, 2013.

In [ ]:
# INSERT_YOUR_ANSWER
#creating a new column 'speed' 
flights['speed'] = flights['distance'] / flights['air_time']
#selecting the fastest flight by sorting
fastest_flight = flights.sort_values('speed', ascending=False).head(1)
#new DataFrame with the information related to the fastest_flight only
exp_res = fastest_flight[['year', 'month', 'day', 'dep_time', 'arr_time', 'origin', 'dest', 'carrier', 'flight']]
print(exp_res)

        year  month  day  dep_time  arr_time origin dest carrier  flight
216447  2013      5   25    1709.0    1923.0    LGA  ATL      DL    1499


---
## Summarize / Aggregate by group
- Pandas: `groupby()` followed by `agg()`

In [ ]:
flights.groupby('month').agg({'dep_delay': ['mean']})

dep_delay
            mean
month           
1      10.036665
2      10.816843
3      13.227076
4      13.938038
5      12.986859
6      20.846332
7      21.727787
8      12.611040
9       6.722476
10      6.243988
11      5.435362
12     16.576688

---
## &diams; Exercise
Which carrier had the worst arrival delays over the year?

(1) filter out all canceled flights; 

(2) group by carrier; next calculate average arrival delay per carrier (You can use `reset_index()` so the grouped result won't have a hierarchical index);

(3) rank your results.

The expected answer is the carrier 'F9' with average arrival delay of 21.920705.

In [ ]:
# ANSWER
# filter out all canceled flights
flights_filtered = flights[flights['arr_delay'].notna()]

# group by carrier; next calculate average arrival delay per carrier using reset_index()
carrier_delay = flights_filtered.groupby('carrier')['arr_delay'].mean().reset_index()

# rank the results
carrier_delay_rank = carrier_delay.sort_values('arr_delay', ascending=False)

# the carrier with worst delay over year
print("The carrier with the worst arrival delay over year is {} with average arrival delay of {}".format(carrier_delay_rank.iloc[0]['carrier'], round(carrier_delay_rank.iloc[0]['arr_delay'],6 )))



The carrier with the worst arrival delay over year is F9 with average arrival delay of 21.920705
